grpc/src/server.rs

```diff
@@ -55,29 +56,31 @@ impl CommunicationsServer for GRPCCommunicationsServer {
         let my_cli_connection =
             GRPCCliConnection::new(self.agent_senders.clone(), self.sender.clone());
 
-        // [impl->swdd~grpc-server-spawns-tonic-service~1]
-        // [impl->swdd~grpc-delegate-workflow-to-external-library~1]
-        let grpc_task = tokio::spawn(async move {
-            Server::builder()
+        let agent_senders_clone = self.agent_senders.clone();
+
+        tokio::select! {
+            // [impl->swdd~grpc-server-spawns-tonic-service~1]
+            // [impl->swdd~grpc-delegate-workflow-to-external-library~1]
+            result = Server::builder()
                 .add_service(AgentConnectionServer::new(my_connection))
                 // [impl->swdd~grpc-server-provides-endpoint-for-cli-connection-handling~1]
                 .add_service(CliConnectionServer::new(my_cli_connection))
-                .serve(addr)
-                .await
-                .map_err(|err| {
-                    CommunicationMiddlewareError(format!(
-                        "Could not start the gRPC service: '{}'",
-                        err
-                    ))
-                })?;
-            Ok::<(), CommunicationMiddlewareError>(())
-        });
-
-        // TODO these two awaits one after the other do not seem correct ...
-        // [impl->swdd~grpc-server-forwards-commands-to-grpc-client~1]
-        execution_command_proxy::forward_from_ankaios_to_proto(&self.agent_senders, receiver).await;
+                .serve(addr) => {
+                    result.map_err(|err| {
+                        GrpcMiddlewareError::StartError(format!("{err:?}"))
```

Comment:
Just one hint: the code produces a debug string as error message which contains nested third party error types (tonic, hyper). So the interesting cause is at the very end. I experimented with `err.source()` when bringing `std::error::Error` into scope, but ended up with the same debug string, because tonic TransportError can have many different issues and not all methods are public to do proper checks for interested sub error types.

Shall we keep the debug err message to have a proper fix for the behavior for now and improve the error message when we introduce generic error handling in general? 

![image](https://github.com/eclipse-ankaios/ankaios/assets/42932060/d086c964-899b-4edd-9d0d-f529fceeb72e)

Reply:
Well, the communication middleware is an exchangeable glue between the server and the agents. In this case the server could not start ("Server startup error") because of the communication middleware ("Communication middleware failure") which in this case is a gRPC middleware ("StartError") which could not start because of a library that is uses.
The current error is maybe long, but it gives some flexibility to exchange parts of the system without too much coupling. The error also is expressive enough as it states where the problem was (although giving too much insides into what libraries we use, which is not a secret).

We could start prescribing how communication middleware error should look like, but we currently support only one and there is a high change that whatever we prescribe will need to be changed later.
As the error message is only logged to the used for reference, I don't see currently the need in any complicated handling here.